[Profiler] Memory profiler part 1: Gradient identification

robieta opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #88653
* #87568
* #87567
* #87566
* #87006
* #86880
* #86854
* #86853
* __->__ #86802
* #87825
* #87133

There are multiple ways to indentify that a Tensor is a gradient. (A subset of which also give additional context.) So to start off I've made a utility to handle that determination.

Differential Revision: [D39920730](https://our.internmc.facebook.com/intern/diff/D39920730/)

Commit:
[Profiler] Memory profiler part 1: Gradient identification

There are multiple ways to indentify that a Tensor is a gradient. (A subset of which also give additional context.) So to start off I've made a utility to handle that determination.

Differential Revision: [D39920730](https://our.internmc.facebook.com/intern/diff/D39920730/)

[ghstack-poisoned]

Commit:
Update on "[Profiler] Memory profiler part 1: Gradient identification"


There are multiple ways to indentify that a Tensor is a gradient. (A subset of which also give additional context.) So to start off I've made a utility to handle that determination.

Differential Revision: [D39920730](https://our.internmc.facebook.com/intern/diff/D39920730/)

[ghstack-poisoned]

Commit:
Update on "[Profiler] Memory profiler part 1: Gradient identification"


There are multiple ways to indentify that a Tensor is a gradient. (A subset of which also give additional context.) So to start off I've made a utility to handle that determination.

Differential Revision: [D39920730](https://our.internmc.facebook.com/intern/diff/D39920730/)

[ghstack-poisoned]

Commit:
Update on "[Profiler] Memory profiler part 1: Gradient identification"


There are multiple ways to indentify that a Tensor is a gradient. (A subset of which also give additional context.) So to start off I've made a utility to handle that determination.

Differential Revision: [D39920730](https://our.internmc.facebook.com/intern/diff/D39920730/)

[ghstack-poisoned]

Commit:
Update on "[Profiler] Memory profiler part 1: Gradient identification"


There are multiple ways to indentify that a Tensor is a gradient. (A subset of which also give additional context.) So to start off I've made a utility to handle that determination.

Differential Revision: [D39920730](https://our.internmc.facebook.com/intern/diff/D39920730/)

[ghstack-poisoned]

robieta commented:
In the interest of keeping code quality high I've opted memory profiler into `mypy-strict`

Commit:
Update on "[Profiler] Memory profiler part 1: Gradient identification"


There are multiple ways to indentify that a Tensor is a gradient. (A subset of which also give additional context.) So to start off I've made a utility to handle that determination.

Differential Revision: [D39920730](https://our.internmc.facebook.com/intern/diff/D39920730/)

[ghstack-poisoned]

Commit:
Update on "[Profiler] Memory profiler part 1: Gradient identification"


There are multiple ways to indentify that a Tensor is a gradient. (A subset of which also give additional context.) So to start off I've made a utility to handle that determination.

Differential Revision: [D39920730](https://our.internmc.facebook.com/intern/diff/D39920730/)

[ghstack-poisoned]

Commit:
Update on "[Profiler] Memory profiler part 1: Gradient identification"


There are multiple ways to indentify that a Tensor is a gradient. (A subset of which also give additional context.) So to start off I've made a utility to handle that determination.

Differential Revision: [D39920730](https://our.internmc.facebook.com/intern/diff/D39920730/)

[ghstack-poisoned]

Commit:
Update on "[Profiler] Memory profiler part 1: Gradient identification"


There are multiple ways to indentify that a Tensor is a gradient. (A subset of which also give additional context.) So to start off I've made a utility to handle that determination.

Differential Revision: [D39920730](https://our.internmc.facebook.com/intern/diff/D39920730/)

[ghstack-poisoned]

Commit:
Update on "[Profiler] Memory profiler part 1: Gradient identification"


There are multiple ways to indentify that a Tensor is a gradient. (A subset of which also give additional context.) So to start off I've made a utility to handle that determination.

Differential Revision: [D39920730](https://our.internmc.facebook.com/intern/diff/D39920730/)

[ghstack-poisoned]

Commit:
Update on "[Profiler] Memory profiler part 1: Gradient identification"


There are multiple ways to indentify that a Tensor is a gradient. (A subset of which also give additional context.) So to start off I've made a utility to handle that determination.

Differential Revision: [D39920730](https://our.internmc.facebook.com/intern/diff/D39920730/)

[ghstack-poisoned]

Commit:
Update on "[Profiler] Memory profiler part 1: Gradient identification"


There are multiple ways to indentify that a Tensor is a gradient. (A subset of which also give additional context.) So to start off I've made a utility to handle that determination.

Differential Revision: [D39920730](https://our.internmc.facebook.com/intern/diff/D39920730/)

[ghstack-poisoned]

Commit:
Update on "[Profiler] Memory profiler part 1: Gradient identification"


There are multiple ways to indentify that a Tensor is a gradient. (A subset of which also give additional context.) So to start off I've made a utility to handle that determination.

Differential Revision: [D39920730](https://our.internmc.facebook.com/intern/diff/D39920730/)

[ghstack-poisoned]

Commit:
Update on "[Profiler] Memory profiler part 1: Gradient identification"


There are multiple ways to indentify that a Tensor is a gradient. (A subset of which also give additional context.) So to start off I've made a utility to handle that determination.

Differential Revision: [D39920730](https://our.internmc.facebook.com/intern/diff/D39920730/)

[ghstack-poisoned]

Commit:
Update on "[Profiler] Memory profiler part 1: Gradient identification"


There are multiple ways to indentify that a Tensor is a gradient. (A subset of which also give additional context.) So to start off I've made a utility to handle that determination.

Differential Revision: [D39920730](https://our.internmc.facebook.com/intern/diff/D39920730/)

[ghstack-poisoned]

Commit:
Update on "[Profiler] Memory profiler part 1: Gradient identification"


There are multiple ways to indentify that a Tensor is a gradient. (A subset of which also give additional context.) So to start off I've made a utility to handle that determination.

Differential Revision: [D39920730](https://our.internmc.facebook.com/intern/diff/D39920730/)

[ghstack-poisoned]

Commit:
Update on "[Profiler] Memory profiler part 1: Gradient identification"


There are multiple ways to indentify that a Tensor is a gradient. (A subset of which also give additional context.) So to start off I've made a utility to handle that determination.

Differential Revision: [D39920730](https://our.internmc.facebook.com/intern/diff/D39920730/)

[ghstack-poisoned]

Commit:
Update on "[Profiler] Memory profiler part 1: Gradient identification"


There are multiple ways to indentify that a Tensor is a gradient. (A subset of which also give additional context.) So to start off I've made a utility to handle that determination.

Differential Revision: [D39920730](https://our.internmc.facebook.com/intern/diff/D39920730/)

[ghstack-poisoned]

Commit:
Update on "[Profiler] Memory profiler part 1: Gradient identification"


There are multiple ways to indentify that a Tensor is a gradient. (A subset of which also give additional context.) So to start off I've made a utility to handle that determination.

Differential Revision: [D39920730](https://our.internmc.facebook.com/intern/diff/D39920730/)

[ghstack-poisoned]

Commit:
Update on "[Profiler] Memory profiler part 1: Gradient identification"


There are multiple ways to indentify that a Tensor is a gradient. (A subset of which also give additional context.) So to start off I've made a utility to handle that determination.

Differential Revision: [D39920730](https://our.internmc.facebook.com/intern/diff/D39920730/)

[ghstack-poisoned]

Commit:
Update on "[Profiler] Memory profiler part 1: Gradient identification"


There are multiple ways to indentify that a Tensor is a gradient. (A subset of which also give additional context.) So to start off I've made a utility to handle that determination.

Differential Revision: [D39920730](https://our.internmc.facebook.com/intern/diff/D39920730/)

[ghstack-poisoned]

albanD reviewed (commented):
I'm not sure about the high level strategy here? What is the exact goal?
If you were recording while the user code was executed, why not just tag all the Tensors created during the backward pass as they get created?

Commit:
Update on "[Profiler] Memory profiler part 1: Gradient identification"


There are multiple ways to indentify that a Tensor is a gradient. (A subset of which also give additional context.) So to start off I've made a utility to handle that determination.

Differential Revision: [D39920730](https://our.internmc.facebook.com/intern/diff/D39920730/)

[ghstack-poisoned]

robieta commented:
> I'm not sure about the high level strategy here? What is the exact goal? If you were recording while the user code was executed, why not just tag all the Tensors created during the backward pass as they get created?

The goal is to distinguish between a proper gradient and a Tensor which is merely created as an implementation detail of an autograd formula. (Since they will generally have different lifetime constraints and implications on roofline memory.) If you recall, one question that came up from some of the plots in the prototype was "why is gradient memory so spiky?" and the answer was that I was tagging all tensors created in the backward pass as gradients rather than the more precise definition that I'm going for here. Does that seem reasonable?

albanD commented:
Ho that make sense. I think the name is a big confusing to me then because all these intermediary that get created are also gradients :p 
The difference is that some of them are the ones requested by the user and some of them are just intermediary results. It might be good to make that distinction in the doc there.

robieta commented:
> Ho that make sense. I think the name is a big confusing to me then because all these intermediary that get created are also gradients :p The difference is that some of them are the ones requested by the user and some of them are just intermediary results. It might be good to make that distinction in the doc there.

I should note that I'm very much open to bikeshedding on the taxonomy (might even be a good composability topic...) and it's very possible that we will end up with a very fine grained internal representation and a more coarse grained set of user facing labels. (Or a switch to decide if some similar categories should be grouped.) Regardless, I'll definitely add a big 'ol docstring to the `Category` Enum.

Commit:
Update on "[Profiler] Memory profiler part 1: Gradient identification"


There are multiple ways to indentify that a Tensor is a gradient. (A subset of which also give additional context.) So to start off I've made a utility to handle that determination.

Differential Revision: [D39920730](https://our.internmc.facebook.com/intern/diff/D39920730/)

[ghstack-poisoned]

Commit:
Update on "[Profiler] Memory profiler part 1: Gradient identification"


There are multiple ways to indentify that a Tensor is a gradient. (A subset of which also give additional context.) So to start off I've made a utility to handle that determination.

Differential Revision: [D39920730](https://our.internmc.facebook.com/intern/diff/D39920730/)

[ghstack-poisoned]

Commit:
Update on "[Profiler] Memory profiler part 1: Gradient identification"


There are multiple ways to indentify that a Tensor is a gradient. (A subset of which also give additional context.) So to start off I've made a utility to handle that determination.

Differential Revision: [D39920730](https://our.internmc.facebook.com/intern/diff/D39920730/)

[ghstack-poisoned]

robieta commented on torch/profiler/_memory_profiler.py:
```diff
@@ -0,0 +1,113 @@
+    children = node.children
+
+    # AccumulateGrad is used in the Autograd engine to handle gradient updates.
+    # There are two possible cases:
```
I added `aten::add` for the double backward case. It's also worth noting that this doesn't have to be 100% foolproof. I mostly just want a reasonable fallback but I'm expecting most gradients to be scooped up by `nn.Module` instrumentation. If you think I've missed any important cases do let me know though.

robieta commented:
@pytorchbot merge
